File: src/main/scala/mrtjp/fengine/assemble/ICFlatMap.scala
package mrtjp.fengine.assemble

import mrtjp.fengine.simulate.{ICGate, ICRegister}
import mrtjp.fengine.tiles.FETileMap

import scala.collection.Map
import scala.collection.Seq
import scala.collection.mutable.ArrayBuffer

trait ICFlatMap
{
    val registers:Map[Int, ICRegister] //The registers currently in the circuit, indexed by ID
    val gates:Map[Int, ICGate] //The gates in the circuit indexed by ID

    val regDependents:Map[Int, ArrayBuffer[Int]]    //[regID -> Seq[gateID]]
    val regDependencies:Map[Int, ArrayBuffer[Int]]  //[regID -> Seq[gateID]]
    val gateDependents:Map[Int, ArrayBuffer[Int]]   //[gateID -> Seq[regID]]
    val gateDependencies:Map[Int, ArrayBuffer[Int]] //[gateID -> Seq[regID]]

    val exploredTileMaps:Seq[FETileMap]
    val exploredFlatMaps:Seq[ICFlatMap]
}

File: src/main/java/mrtjp/fengine/api/ICFlatMap.java
package mrtjp.fengine.api;

import mrtjp.fengine.simulate.ICGate;
import mrtjp.fengine.simulate.ICRegister;
import mrtjp.fengine.tiles.FETileMap;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ICFlatMap {

    private final Map<Integer, ICRegister> registers;
    private final Map<Integer, ICGate> gates;
    private final Map<Integer, ArrayList<Integer>> registerDependents;
    private final Map<Integer, ArrayList<Integer>> registerDependencies;
    private final Map<Integer, ArrayList<Integer>> gateDependents;
    private final Map<Integer, ArrayList<Integer>> gateDependencies;

    public ICFlatMap(Map<Integer, ICRegister> registers, Map<Integer, ICGate> gates, Map<Integer, ArrayList<Integer>> registerDependents, Map<Integer, ArrayList<Integer>> registerDependencies, Map<Integer, ArrayList<Integer>> gateDependents, Map<Integer, ArrayList<Integer>> gateDependencies) {
        this.registers = registers;
        this.gates = gates;
        this.registerDependents = registerDependents;
        this.registerDependencies = registerDependencies;
        this.gateDependents = gateDependents;
        this.gateDependencies = gateDependencies;
    }

    public Map<Integer, ICRegister> getRegisters() { return registers; }
    public Map<Integer, ICGate> getGates() { return gates; }
    public Map<Integer, ArrayList<Integer>> getRegDependents() { return registerDependents; }    //[regID -> Seq[gateID]]
    public Map<Integer, ArrayList<Integer>> getRegDependencies() { return registerDependencies; }  //[regID -> Seq[gateID]]
    public Map<Integer, ArrayList<Integer>> getGateDependents() { return gateDependents; }   //[gateID -> Seq[regID]]
    public Map<Integer, ArrayList<Integer>> getGateDependencies() { return gateDependencies; } //[gateID -> Seq[regID]]
}
